### DISCUSSION TITLE
dependency on glog (via txtpbfmt) adds some flags to a program's std flag list

### DISCUSSION THREAD

**Discussion by knieriem** [cue-lang/cue#1196] [General]
When upgrading a program importing a package from cue v0.3.0-beta* to v0.4.0 I noted that CUE's new dependency github.com/protocolbuffers/txtpbfmt adds another dependency, github.com/golang/glog (leveled execution logs for Go). `go mod why` lists, for instance:
```
cuelang.org/go/cue/load
  cuelang.org/go/internal/encoding
    cuelang.org/go/encoding/protobuf/textproto
       github.com/protocolbuffers/txtpbfmt/parser
         github.com/golang/glog
```

Glog, as a package, [in its init function defines some command line flags][glogflags] using Go's standard flag package:

    -alsologtostderr
    -log_backtrace_at value
    -log_dir string
    -logtostderr
    -stderrthreshold value
    -v value
    -vmodule value

[glogflags]: https://github.com/golang/glog/blob/424d2337a5299a465c8a8228fc3ba4b1c28337a2/glog.go#L398-L404

As a result, any program importing, for example, `cuelang.org/go/cue/load` will automatically inherit these flags. In my program, which makes use of the standard flag package, I had been defining an own `flag.Uint("v", 0, "verbosity level")`.
With cue v0.4.0 I'm getting a `panic: ... flag redefined: v` error now.

Since the `cue` program itself is using cobra+pflag, this effect is not visible when using cue itself, since flags defined using Go's standard flag package are ignored then.

An old thread at golang-nuts, [_glog and testing flag "pollution"_](https://groups.google.com/g/golang-nuts/c/vj8ozVqemnQ) mentions a workaround: defining one's own flag.FlagSet instead of using the default `flag.CommandLine`. But this also means, that all flags defined by a program's internal packages using e.g. `flag.Bool` must be adapted to the new FlagSet.
An alternative to this workaround would be to make `textproto` depend on a modified `txtpbfmt` that does not depend on _glog_. I think for my use-case I will apply the method described in that golang-nuts thread.

**Comment by ChrisHines** [comment]
Another approach is to replace `glog` with an API equivalent that redirects the `glog` calls to something else and removes the global flags added by `glog` as a result. I've seen several packages that replace `glog` with shims for different specific logging packages. https://github.com/slok/noglog shows how the technique works in a general way.

**Reply by knieriem** [reply]
Thanks for this hint. Adding

     replace github.com/golang/glog => github.com/slok/noglog v0.2.0 

to `go.mod` appears to be sufficient. There remains a problem when running `go mod tidy`, which fails because the `module` directive of noglog's go.mod file does not match glog's (as required by the [modules reference][mod]), but this could be fixed.

[mod]: https://golang.org/ref/mod#go-mod-file-replace

**Reply by myitcv** [reply]
Using a `replace` directive is, however, only a temporary solution, because it creates downstream problems of its own. Hence why this needs to be fixed upstream; I wouldn't want the CUE project to be offering this as a permanent fix suggestion to users if the "fix" creates more problems for them.

**Comment by myitcv** [comment]
This seems rather unfortunate to put it mildly, and something we should attempt to fix via a PR to the `github.com/protocolbuffers/txtpbfmt` module. Thanks for raising, @knieriem - I will create an issue tracking that off the back of this discussion.

**Reply by gurjeet** [reply]
The `good first issue` label made this problem look deceptively simple; false advertising :-)

But I have take a crack at solving this problem, so if the fix is merged, all users of the `glog` package will be able to override these flags at build time. See https://github.com/golang/glog/pull/54
